### PR TITLE
WIP: Lifecycle hooks consistency & async scheduler tasks

### PR DESCRIPTION
**Pull request by fkleuver**
# Pull Request

## 📖 Description

In this PR:
- Additional lifecycle hooks added on specific timings, and minor changes to the timings and order of some hooks. Purpose is to improve consistency and make them more compatible with scheduling, time-slicing, task cancellation and animation.
  - Combined `bind` and `attach` APIs on the controller interface into `activate`
  - Combined `detach` and `unbind` APIs into `deactivate`
  - In the built-in resources, `activate` is called during `afterAttach` and `deactivate` is called during `afterUnbind`. This means that, rather than lifecycle invocations going phase-by-phase (first all `bind`, then all `attach`, etc), they now go controller-by-controller. This allows controllers to be scheduled and time-sliced, makes first-class support for render-blocking animations and suspense (in cleaner forms) possible, simplifies integration with the router and generally reduces the "attack surface", or more specifically, the ways in which the API can be used wrongly.
  - All lifecycle hooks are now awaited if the component returns a promise. This means:
    - When returning a promise from `afterAttach` (that resolves after e.g. an animation is done), this will block activation of child components until the promise resolves. The same already applied for binding hooks
    - When returning a promise from `beforeDetach`, this will block unmounting of the component itself as well as deactivation of child components until the promise resolves (likewise, the same already applied for unbinding hooks)
  - `afterBind` changed to the timing of "after binding self, but before binding children"
  - `afterUnbindChildren` added at the original timing of `afterUnbind`
  - `afterUnbind` changed to the timing of "after unbinding self, but before unbinding children"
  - unbind lifecycle changed from depth-first to breadth-first to make the above possible and to be consistent with the bind lifecycle
  - `afterAttachChildren` added at the original timing of `afterAttach`
  - `afterAttach` added changed to the timing of "after attaching self, but before attaching children"
  - removed `beforeAttach` because it has no use anymore (timing would be identical to `afterBind`)
  - removed `afterDetach` because it has no use anymore (timing would be identical to `beforeUnbind`)

  - `dispose` added which is called after `afterUnbindChildren` (the last "out" hook), if and only if `.release()` is called on the controller (or one of its parents) prior to deactivation, and no parent view has caching enabled (in that case that view along with its whole subtree is returned to cache instead).

- Added an async task api to the scheduler, for tasks that need to be awaited internally. This will be useful for, among others:
  - Realizing otherwise difficult scenarios such as animating via async attach/detach hooks with time-slicing enabled, etc.
  - Simplifying the wireup of router lifecycle hooks
  - Removing the `ILifecycleTask` interface and its implementations, significantly simplifying the runtime

### CODE REVIEW

**Comment by bigopon** [comment]
Just a friendly reminder that this PR should go in together with the doc related to lifecycle is updated. Also, should be include the signature of the call in the doc?

**Comment by fkleuver** [comment]
> Just a friendly reminder that this PR should go in together with the doc related to lifecycle is updated.
Which file is that?

> Also, should be include the signature of the call in the doc?
Preferably not yet. I want to try to get rid of those flags in the signature, so adding them to the docs now would just be more work later to change it back again

**Comment by bigopon** [comment]
It's here https://docs.aurelia.io/getting-started/components#the-component-lifecycle

If you asked about where the file is, it's here https://github.com/aurelia/aurelia/blob/master/docs/user-docs/getting-started/components.md#the-component-lifecycle

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://45297-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://46397-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

File: packages/runtime-html/src/resources/custom-attributes/blur.ts [hunk 1]
```diff
@@ -157,7 +157,7 @@ export class Blur implements ICustomAttributeViewModel<HTMLElement> {
     this.manager = BlurManager.createFor(dom, scheduler);
   }
 
-  public afterAttach(): void {
+  public afterAttachChildren(): void {
```
Comment (bigopon) [inline packages/runtime-html/src/resources/custom-attributes/blur.ts:160]: this custom attribute doesn't need to wait for the children of its host. `afterAttach` should be ok, if I understand the timing correctly

Comment (fkleuver) [reply packages/runtime-html/src/resources/custom-attributes/blur.ts:160]: The `afterAttachChildren` hook is identical to the timing of the hook it was previously (`afterAttach`). You sure though? Doesn't blur need to wait for everything to be on the page?

Comment (bigopon) [reply packages/runtime-html/src/resources/custom-attributes/blur.ts:160]: The attribute itself reacts to user action, so it doesn't need to wait for any of its children. And ideally should be ready to go asap

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://46406-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://47838-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49341-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49368-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49433-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49471-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49477-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49523-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49603-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49624-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://49843-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50085-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50111-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50131-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50137-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50185-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50232-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50260-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50386-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50483-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50608-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50696-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://50922-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://51204-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://51212-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://51219-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://51229-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://52456-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://53095-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://53544-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://53553-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://53585-121594173-gh.circle-artifacts.com/0/test/js-framework-benchmark/webdriver-ts-results/table.html

**Pull request closed without merging**